### PR TITLE
#46 Increases min gas price to 500000000000 and fixes ABCI sad path.

### PR DESCRIPTION
**Pull request by EnchanterIO**
* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests

### CODE REVIEW

File: consensus/abci.go [hunk 1]
```diff
@@ -123,17 +122,50 @@ func (abci *TendermintABCI) CheckTx(txBytes []byte) tmtAbciTypes.ResponseCheckTx
 	
 	tx, err := decodeRLP(txBytes)
 	if err != nil {
-		abci.logger.Error("Received invalid transaction", "tx", tx.Hash().String())
-		abci.metrics.CheckErrTxsTotal.Add(1, fmt.Sprint(abciTypes.ErrEncodingError.Code))
-		return tmtAbciTypes.ResponseCheckTx{Code: uint32(abciTypes.ErrEncodingError.Code), Log: err.Error()}
+		abci.logger.Error("Unable to decode RLP TX", "err", err.Error())
+		abci.metrics.CheckErrTxsTotal.Add(1, "INVALID_TX")
```
Comment (ggarri) [inline consensus/abci.go]: [MINOR] Please encapsulate every error msg code into its own `error.go` file so we can identify where they have been used.

File: consensus/abci.go [hunk 2]
```diff
@@ -123,17 +122,50 @@ func (abci *TendermintABCI) CheckTx(txBytes []byte) tmtAbciTypes.ResponseCheckTx
...
+
+	return tmtAbciTypes.ResponseCheckTx{Code: tmtAbciTypes.CodeTypeOK}
+}
 
+func (abci *TendermintABCI) doMempoolValidation(tx *ethTypes.Transaction) (address common.Address, err error) {
 	if tx.Size() > maxTransactionSize {
-		abci.logger.Error(core.ErrOversizedData.Error())
-		abci.metrics.CheckErrTxsTotal.Add(1, fmt.Sprint(abciTypes.ErrInternalError.Code))
-		return tmtAbciTypes.ResponseCheckTx{Code: uint32(abciTypes.ErrInternalError.Code), Log: core.ErrOversizedData.Error()}
+		return common.Address{}, fmt.Errorf("MAX_TRANSACTION_SIZE")
```
Comment (ggarri) [inline consensus/abci.go]: [MINOR] Encapsulate msg error into a constant somewhere else

File: consensus/abci.go [hunk 3]
```diff
@@ -143,63 +175,32 @@ func (abci *TendermintABCI) CheckTx(txBytes []byte) tmtAbciTypes.ResponseCheckTx
 
 	from, err := ethTypes.Sender(signer, tx)
 	if err != nil {
-		abci.logger.Error(core.ErrInvalidSender.Error())
-		abci.metrics.CheckErrTxsTotal.Add(1, fmt.Sprint(abciTypes.ErrBaseInvalidSignature.Code))
-		return tmtAbciTypes.ResponseCheckTx{Code: uint32(abciTypes.ErrBaseInvalidSignature.Code), Log: core.ErrInvalidSender.Error()}
+		return common.Address{}, fmt.Errorf("INVALID_SENDER")
```
Comment (ggarri) [inline consensus/abci.go]: [MINOR] Encapsulate msg error into a constant somewhere else

File: consensus/abci.go [hunk 4]
```diff
@@ -143,63 +175,32 @@ func (abci *TendermintABCI) CheckTx(txBytes []byte) tmtAbciTypes.ResponseCheckTx
...
 	}
 
 	if tx.Value().Sign() < 0 {
-		abci.logger.Error(core.ErrNegativeValue.Error())
-		abci.metrics.CheckErrTxsTotal.Add(1, fmt.Sprint(abciTypes.ErrBaseInvalidInput.Code))
-		return tmtAbciTypes.ResponseCheckTx{Code: uint32(abciTypes.ErrBaseInvalidInput.Code), Log: core.ErrNegativeValue.Error()}
+		return common.Address{}, fmt.Errorf("INVALID_SIGNATURE")
```
Comment (ggarri) [inline consensus/abci.go]: [MINOR] Encapsulate msg error into a constant somewhere else

File: consensus/abci.go [hunk 5]
```diff
@@ -123,17 +122,50 @@ func (abci *TendermintABCI) CheckTx(txBytes []byte) tmtAbciTypes.ResponseCheckTx
...
+	}
+
+	abci.checkTxState.SetNonce(from, tx.Nonce() + 1)
+
+	abci.logger.Info("TX validated.", "hash", tx.Hash().String(), "state_nonce", abci.checkTxState.GetNonce(from))
+
+	return tmtAbciTypes.ResponseCheckTx{Code: tmtAbciTypes.CodeTypeOK}
+}
 
+func (abci *TendermintABCI) doMempoolValidation(tx *ethTypes.Transaction) (address common.Address, err error) {
```
Comment (ggarri) [inline consensus/abci.go]: [MAJOR] I cannot understand why a method responsible for doing a validation returns a relevant value such as `from` account back. You clearly break the single responsibility propose of this method.

File: consensus/abci.go [hunk 6]
```diff
@@ -123,17 +122,50 @@ func (abci *TendermintABCI) CheckTx(txBytes []byte) tmtAbciTypes.ResponseCheckTx
...
+		abci.logger.Error("TX gas price too low", "tx_gas_price", tx.GasPrice(), "required_tx_gas_price", abci.db.Config().TxPool.PriceLimit)
+		abci.metrics.CheckErrTxsTotal.Add(1, "INSUFFICIENT_GAS_PRICE")
+		return tmtAbciTypes.ResponseCheckTx{Code: 1, Log: "INSUFFICIENT_GAS_PRICE"}
+	}
+
+	currentBalance := abci.checkTxState.GetBalance(from)
+	txCost := tx.Cost()
+	if currentBalance.Cmp(txCost) < 0 {
+		abci.logger.Error("Low balance", "balance", currentBalance.String(), "tx_cost", tx.Cost().String())
+		abci.metrics.CheckErrTxsTotal.Add(1, "INSUFFICIENT_FUNDS")
```
Comment (ggarri) [inline consensus/abci.go]: [MINOR] Refactor msg error string

File: consensus/abci.go [hunk 7]
```diff
@@ -123,17 +122,50 @@ func (abci *TendermintABCI) CheckTx(txBytes []byte) tmtAbciTypes.ResponseCheckTx
...
-		abci.logger.Error("Received invalid transaction", "tx", tx.Hash().String())
-		abci.metrics.CheckErrTxsTotal.Add(1, fmt.Sprint(abciTypes.ErrEncodingError.Code))
-		return tmtAbciTypes.ResponseCheckTx{Code: uint32(abciTypes.ErrEncodingError.Code), Log: err.Error()}
+		abci.logger.Error("Unable to decode RLP TX", "err", err.Error())
+		abci.metrics.CheckErrTxsTotal.Add(1, "INVALID_TX")
+		return tmtAbciTypes.ResponseCheckTx{Code: 1, Log: "INVALID_TX"}
 	}
 
 	abci.logger.Info("Checking TX", "hash", tx.Hash().String(), "nonce", tx.Nonce(), "cost", tx.Cost())
+	from, err := abci.doMempoolValidation(tx)
```
Comment (ggarri) [inline consensus/abci.go]: [MAJOR] Why a `doMempoolValidation` returns an error along with a Eth account. That is hard to understand an it breaks the Single Responsibility propose of this method.

File: consensus/abci.go [hunk 8]
```diff
@@ -143,63 +175,32 @@ func (abci *TendermintABCI) CheckTx(txBytes []byte) tmtAbciTypes.ResponseCheckTx
...
-	}
-
-	currentBalance := abci.checkTxState.GetBalance(from)
-	txCost := tx.Cost()
-	if currentBalance.Cmp(txCost) < 0 {
-		errMsg := fmt.Sprintf("Current balance: %s, TX cost: %s", currentBalance, tx.Cost())
-		abci.logger.Error(errMsg)
-		abci.metrics.CheckErrTxsTotal.Add(1, fmt.Sprint(abciTypes.ErrInsufficientFunds.Code))
-		return tmtAbciTypes.ResponseCheckTx{Code: uint32(abciTypes.ErrInsufficientFunds.Code), Log: errMsg}
+		return common.Address{}, fmt.Errorf("INTRINSIC_GAS")
```
Comment (ggarri) [inline consensus/abci.go]: [MINOR] Wrong message. Maybe "NOT_ENOUGH_INTRINSIC_GAS"

File: consensus/abci.go [hunk 9]
```diff
@@ -123,17 +122,50 @@ func (abci *TendermintABCI) CheckTx(txBytes []byte) tmtAbciTypes.ResponseCheckTx
...
 	abci.logger.Info("Checking TX", "hash", tx.Hash().String(), "nonce", tx.Nonce(), "cost", tx.Cost())
+	from, err := abci.doMempoolValidation(tx)
+	if err != nil {
+		abci.logger.Error(err.Error())
+		abci.metrics.CheckErrTxsTotal.Add(1, err.Error())
+		return tmtAbciTypes.ResponseCheckTx{Code: 1, Log: err.Error()}
+	}
+
+	// Additional custom validation aside of the Ethereum default mempool validation
+	if tx.GasPrice().Uint64() < abci.db.Config().TxPool.PriceLimit {
```
Comment (ggarri) [inline consensus/abci.go]: [QUESTION] Is this validation adding any additional security over the `IntrinsicGas` validation? 

[MINOR} What is the reason to have this validation as not part of the `doMempoolValidation`? As I see it, we are checking the gas twice in two manners and there is not sense to be splitting that validation into two different places, that will add confusion.

File: database/database.go [hunk 10]
```diff
@@ -178,6 +171,19 @@ func (db *Database) Protocols() []p2p.Protocol {
 	return nil
 }
 
+func (db *Database) UpdateExecuteTxMetrics(tx *ethTypes.Transaction) {
```
Comment (ggarri) [inline database/database.go]: [MINOR] I cannot see reason to encapsulate this code on a separate method when there is only one usage of it in the whole code. This encapsulate brings more complexity and not reusability. I would rather mode this code back to `ExecuteTx` unless you have already an usage in mind for having it separated

File: database/config.go [hunk 11]
```diff
@@ -76,6 +79,16 @@ func NewConfig(dataDir string, metrics bool, ctx *cli.Context) (Config, error) {
 	gethCfg.EthCfg.TrieDirtyCache = 0 // MB
 	gethCfg.EthCfg.TrieTimeout = 5 * time.Minute
 
+	// To prevent DDOS TX spam, a min gas price must be defined
```
Comment (ggarri) [inline database/config.go]: As you already gathered the knowledge regarding the meaning behind each of this values. I think it would be a good idea to include a comment over each of them for future references and prevent next developers to waste their time again investigating the reason for each of these assigments

File: database/txclient/tx_config.go [hunk 12]
```diff
@@ -24,6 +27,7 @@ func NewTxConfig(gasLimit uint64, receiptTimeout time.Duration, receiptInterval
 func NewTransferTxConfig() TxConfig {
```
Comment (ggarri) [inline database/txclient/tx_config.go]: [MINOR] It would reduce duplicity if you would have written:
```
return NewTxConfig(database.MinGasPrice, 21000, TxReceiptTimeout...)
```

File: database/web3/currency.go [hunk 13]
```diff
@@ -4,19 +4,20 @@ import (
 	"math/big"
 	"fmt"
 	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/common/math"
 )
 
-func PhotonToWei(value string) (*big.Int, error) {
```
Comment (ggarri) [inline database/web3/currency.go]: [MINOR] Why did you remove this method? Does it not look useful? Having us to calculate amounts in Wei it is adding alot of overhead, I would suggest to restore this method unless there is a better alternative to convert PHT to WEI

File: truffle/test/03_smart_contract.js [hunk 14]
```diff
@@ -60,15 +61,23 @@ describe('SmartContract', async () => {
...
 
     try {
       const tx = await instance.incrementHelloCount({
-        from: ROOT_ACCOUNT
+        from: ROOT_ACCOUNT,
+        gas: 100000, // Shouldn't take more than 100 000 to perform a TX
       });
       const txReceipt = await fetchTxReceipt(tx.tx);
+      const weiBNBalancePostTx = await web3.eth.getBalance(ROOT_ACCOUNT);
+      const expectedGasCost = calculateGasCostBN(txReceipt.gasUsed);
```
Comment (ggarri) [inline truffle/test/03_smart_contract.js]: [MINOR] You cannot call this `expected` when you are retrieving the right value from the txReceipt information. That is very confusing. Renamed to just `txGasCost` to avoid future missunderstandings

File: cmd/lightchain/init.go [hunk 15]
```diff
@@ -72,24 +77,41 @@ func newNodeCfgFromCmd(cmd *cobra.Command) (node.Config, setup.Network, error) {
 	dataDir, _ := cmd.Flags().GetString(DataDirFlag.Name)
 	useStandAloneNet, _ := cmd.Flags().GetBool(StandAloneNetFlag.Name)
 	useSiriusNet, _ := cmd.Flags().GetBool(SiriusNetFlag.Name)
+	forceInit, _ := cmd.Flags().GetBool(ForceFlag.Name)
 
 	shouldTrace, _ := cmd.Flags().GetBool(TraceFlag.Name)
 	traceLogFilePath, _ := cmd.Flags().GetString(TraceLogFlag.Name)
 
+	// To prevent accidental data dir like "/" and disasters
```
[the remark this answers is not in the document]
Comment (EnchanterIO) [reply cmd/lightchain/init.go]: or u are root yourself or have bad permissions ;) It's justified u just have to think more broadly. You could also do smth like ~/ and so on. Anyway it was for a peace of mind if we make a bug and make the flag optional or not retrieve the value properly. In that case u could get a default value of "" and remove god knows what.

File: cmd/lightchain/init.go [hunk 16]
```diff
@@ -72,24 +77,41 @@ func newNodeCfgFromCmd(cmd *cobra.Command) (node.Config, setup.Network, error) {
...
 	// This should be done inside of the `node.Init()` pkg but due to bad design,
 	// creation of new Ethereum node instance from a config accidentally modifies the FS by creating
 	// a keystore dir therefore we have to perform this check as early as possible.
 	//
 	// Todo: After #90 is done, move this code to `Node.Init()`
 	if doesExist, err := fs.DirExists(dataDir); doesExist || err != nil {
-		return node.Config{}, "", fmt.Errorf("unable to initialize lightchain node. %s already exists", dataDir)
+		if forceInit {
+			logger.Info(fmt.Sprintf("Forcing '%s' data dir removal...", dataDir))
+			if err := fs.RemoveAll(dataDir); err != nil {
```
[the remark this answers is not in the document]
Comment (EnchanterIO) [reply cmd/lightchain/init.go]: I can create a task for this.

File: cmd/lightchain/main.go [hunk 17]
```diff
@@ -45,6 +46,12 @@ var (
 func main() {
 	lightchainCmd := LightchainCmd()
 
+	defer func() {
```
[the remark this answers is not in the document]
Comment (EnchanterIO) [reply cmd/lightchain/main.go]: it prints additional reason of the panic not available normally in stack trace. i will verify it still exits with 1 code

File: consensus/abci.go (hunk 5 above)
Comment (EnchanterIO) [reply consensus/abci.go]: ahhh ok

File: consensus/abci.go (hunk 1 above)
Comment (EnchanterIO) [reply consensus/abci.go]: u can see where they have been used, only here so no constant needed for now it just makes it harder to maintain, jumping through codebase.

File: consensus/abci.go (hunk 2 above)
Comment (EnchanterIO) [reply consensus/abci.go]: u can see where they have been used, only here so no constant needed for now it just makes it harder to maintain, jumping through codebase.

File: consensus/abci.go (hunk 3 above)
Comment (EnchanterIO) [reply consensus/abci.go]: u can see where they have been used, only here so no constant needed for now it just makes it harder to maintain, jumping through codebase.

File: consensus/abci.go (hunk 4 above)
Comment (EnchanterIO) [reply consensus/abci.go]: u can see where they have been used, only here so no constant needed for now it just makes it harder to maintain, jumping through codebase.

File: consensus/abci.go (hunk 7 above)
Comment (EnchanterIO) [reply consensus/abci.go]: because it was convenient and I was in rush but yea I agree, PS: [MINOR]

File: consensus/abci.go (hunk 8 above)
Comment (EnchanterIO) [reply consensus/abci.go]: that's obvious from the name instrinsic but well, I will change it just for you 💋  INSUFFICIENT_INTRINSIC_GAS

File: consensus/abci.go (hunk 9 above)
Comment (EnchanterIO) [reply consensus/abci.go]: gas amount != gas price, they are different concepts. IntrinsicGas defines the amount of gas necessary for an operation and is constant in case of funds transfers, gas price works like an auction system and is dynamic. Also this validation together with balances is custom and not part of the shared mempool validation

File: database/database.go (hunk 10 above)
Comment (EnchanterIO) [reply database/database.go]: The reason for encapsulating the metrics code is that it's not suppose to be in the important method such as ExecuteTx and poluting it in first place but should have been done inside the metrics package. I also encapsulated it also to keep the SRP of ExecuteTx as it was doing more stuff than just metrics

File: database/config.go (hunk 11 above)
Comment (EnchanterIO) [reply database/config.go]: sure, i will add more explanation for each variable

File: database/txclient/tx_config.go (hunk 12 above)
Comment (EnchanterIO) [reply database/txclient/tx_config.go]: nice idea, done

File: database/web3/currency.go (hunk 13 above)
Comment (EnchanterIO) [reply database/web3/currency.go]: because it was not very useful and it worked only with integers. I replaced it with ParseWei that is more flexible. At the end if Ethereum doesn't handle floats anywhere in the code maybe we shouldn't go that extra mile needed. We can add additional method like PhotonToWEi in the future but only if it will work well with all types, string, integers and floats. For now ParseWei() will do the trick

File: truffle/test/03_smart_contract.js (hunk 14 above)
Comment (EnchanterIO) [reply truffle/test/03_smart_contract.js]: renamed to txGasCost

File: cmd/lightchain/init.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (EnchanterIO) [reply cmd/lightchain/init.go]: https://github.com/lightstreams-network/lightchain/issues/98
